### PR TITLE
Update documentation

### LINKED ISSUE #156: 2.0.0 release
Let's use this issue to keep track of the to-do list for the upcoming 2.0.0 release. Now that the current version lives on its own branch, we should be able to merge #153 into master and then work from there.

To do:
- [x] Update all documentation
- [x] Let the `Settings` class try to infer values from environment variables if present
- [x] Provide a docker image with an entrypoint that can easily be overridden to add custom plugins etc.
- [ ] Add migration guide to docs
- [ ] Improve logging (it would be nice if the bot facilitates easily logging to file, and the logs should be more extensive with messages of various logging levels)

At release time:
- [x] Publish the docker image as `mmpy_bot:2.0.0`.
- [x] Publish new PyPi package `mmpy_bot:2.0.0`

### PR DESCRIPTION
**Pull request by ttuffin**
Update all documentation as required for #156.

TODO:
- [x] `contributing.rst` @ttuffin 
- [x] Update `README.md` @ttuffin 
- [x] remove `decorators.rst` and integrate into `plugins`
- [x] merge `installation.rst`, `deploy.rst` and `usage.rst` into single deployment doc @ttuffin 
- [x] update `dev-install.rst` (@ttuffin is working on this). I have already updated `testing.rst`, but perhaps we should merge these two into a single `development` doc? ttuffin: has been merged into `contributing.rst`
- [x] Update `settings.rst` (I'll take care of this, but I'm not sure we even need it. The `Settings` class is pretty self-explanatory, so perhaps we could just add a link to the source code in the `deploy` section.
- [x] Update `use_cases.rst`, but I'm not a fan of this document at all to be honest. Again, the `ExamplePlugin` is quite self-explanatory, so if we just link to that source code in one of the other docs we should be good.
- [ ] Create `migration.rst` to document 1.x -> 2.x process. @unode  
- [x] Integrate `deploy.rst` into `getting_started.rst` and delete the old file.

### CODE REVIEW

**Comment by ttuffin** [comment]
Started with plugins documentation, still need to update the rest. Feel free to update/change as you see fit.

File: docs/dev_install.rst
Comment (attzonko) [inline docs/dev_install.rst:20]: missing word "set"

File: docs/dev_install.rst
Comment (attzonko) [inline docs/dev_install.rst:11]: Why don't we provide the same command used by the workflow `pip install -e  .[dev]`

File: docs/dev_install.rst
Comment (attzonko) [inline docs/dev_install.rst:4]: Do we want to add documentation on using `venv` module, to isolate the development environment from their default python environment?

File: docs/plugins.rst
Comment (attzonko) [inline docs/plugins.rst:52]: should be "expect `the` appropriate reply."

File: docs/plugins.rst
Comment (attzonko) [inline docs/plugins.rst:85]: We should write a quick blurb here explaining the benefit of using Click?

File: docs/dev_install.rst
Comment (jneeven) [inline docs/dev_install.rst:19]: The idea is to *not* edit settings.py, but just to create a  `Settings` object where you specify the settings you want to change (e.g. like we do for the integration tests). But this will likely need an update after I've implemented the environment variables, so we can leave it like this for now.

File: docs/plugins.rst
Comment (jneeven) [inline docs/plugins.rst:9]: Since a custom bot does not require checking out the source code (I'd even recommend against it), there won't be a plugins directory. Users can just subclass `Plugin` wherever they want and pass it to their `Bot` constructor (as with the settings)

File: docs/plugins.rst
Comment (jneeven) [inline docs/plugins.rst:14]:
```suggestion
        from mmpy_bot import Plugin, listen_to
```

File: docs/plugins.rst
Comment (jneeven) [inline docs/plugins.rst:15]:
```suggestion
        from mmpy_bot import Message
```

File: docs/plugins.rst
Comment (jneeven) [inline docs/plugins.rst:48]: This is not necessary unless they are contributing

File: docs/plugins.rst
Comment (jneeven) [inline docs/plugins.rst:49]: As mentioned above, they don't need to modify any files other than their main file. They can just have a file like this:

```python
from mmpy_bot import Bot, Settings
from my_plugin import MyPlugin

bot = Bot(settings=Settings(BOT_TOKEN="my_token"), plugins=[MyPlugin()])
bot.run()
```

File: docs/plugins.rst
Comment (jneeven) [inline docs/plugins.rst:245]: All of this is still relevant, I've only added the `subprocess` tag

**Review by jneeven — COMMENTED** [review]
Awesome, thanks  for getting this started!

File: docs/plugins.rst
Comment (ttuffin) [reply docs/plugins.rst:49]: Thanks, that makes a lot more sense!

**Comment by ttuffin** [comment]
Great feedback :+1: I will update with the proposed changes later today.

File: docs/plugins.rst
Comment (jneeven) [reply docs/plugins.rst:49]: I've created an `entrypoint.py` in #162 that we can refer to in the docs

**Comment by jneeven** [comment]
I've updated two of the docs and added a to-do list for the remaining ones in https://github.com/attzonko/mmpy_bot/pull/161#issue-592115248. Let me know what you think about those proposals!

File: docs/plugins.rst
Comment (ttuffin) [reply docs/plugins.rst:85]: Agreed, but to be honest, I don't even know myself what it's for :smile:. @jneeven can you write a quick couple of lines about it?

**Comment by ttuffin** [comment]
> I've updated two of the docs and added a to-do list for the remaining ones in [#161 (comment)](https://github.com/attzonko/mmpy_bot/pull/161#issue-592115248). Let me know what you think about those proposals!

Agree with you in regards to dev-install.rst and testing.rst. I am updating contributing.rst to combine both of these topics, since they all fall under the contributor umbrella.

File: README.md
Comment (ttuffin) [inline README.md:65]: Is this still valid? It's from the legacy README.md but I don't think this is the case anymore?

File: docs/plugins.rst
Comment (jneeven) [reply docs/plugins.rst:94]: Click is amazing! I'll write something tomorrow :+1:

File: README.md
Comment (jneeven) [reply README.md:65]: The description is a bit cryptic, but the bot does indeed need system admin permissions to use the entire API, e.g. ephemeral replies.

Comment (ttuffin) [reply README.md]: Cool, I've updated the doc to clarify this a little better.

**Comment by unode** [comment]
Hi everyone, I'm trying to update my existing code and I'm finding a few cases that might benefit from a "migrating from 1.0 to 2.0" kind of document. A few questions so far:

1. Is the "Django-like" `settings.py` completely gone? If I had extra stuff in my `settings.py` do I need to subclass the `Settings` dataclass? 
2. Do I understand correctly that the `click` interface is meant to be used via mattermost and not for command-line invocations?
3. There's now only `listen_to`. I see a `needs_mention=` argument that seems to make it behave like `respond_to`. Is there anything like [`at_start`](https://mmpy-bot.readthedocs.io/en/latest/plugins.html?highlight=at_start#id1) hiding somewhere?

**Comment by jneeven** [comment]
> a "migrating from 1.0 to 2.0" kind of document.

Very good idea!

> If I had extra stuff in my settings.py do I need to subclass the Settings dataclass?

This is indeed the way to do it. I'm not very familiar with Django, have I accidentally prevented any valid use-cases now?

> Do I understand correctly that the click interface is meant to be used via mattermost and not for command-line invocations?

Correct again ;) I found this much more robust and straightforward than messing with regular expressions if you want to execute commands that way. Of course, if you do want to use regular expressions and / or take a more "natural language" kind of approach, that's still supported.

> Is there anything like at_start hiding somewhere?

Yes, [there is indeed](https://github.com/attzonko/mmpy_bot/blob/master/mmpy_bot/plugins/base.py#L60). And an `on_stop` as well, should you need it.

If any more questions arise, please keep asking them! A "FAQ" like this is probably at least as valuable as any documentation updates will be :+1:

**Comment by unode** [comment]
> This is indeed the way to do it. I'm not very familiar with Django, have I accidentally prevented any valid use-cases now?

I think not but I'll confirm this once I get 2.0 running locally with our 1.3 plugins. We had some secrets we were importing from `settings.py` and will likely need a different approach now.

**Comment by ttuffin** [comment]
@unode that's great if you can document the migration from 1.x -> 2.x :+1: 

I'll add it to the to-do list.

**Comment by ttuffin** [comment]
I think `installation.rst`, `deploy.rst` and `usage.rst` can be removed/merged in favour of one deployment document.

- `deploy.rst` is specific to Ubuntu (apt), and we should try to keep instructions distribution agnostic if possible.
- The current `installation.rst` is short and can be included in another document without making it overblown.

In essence, a single document that:
1. Details installation procedure via Git repo, pip, and containers.
2. How to run the bot and define settings and plugins through an `entrypoint` file.
3. A link to `plugins.rst` for writing custom plugins.

**Comment by attzonko** [comment]
> I think `installation.rst`, `deploy.rst` and `usage.rst` can be removed/merged in favour of one deployment document.
> 
> * `deploy.rst` is specific to Ubuntu (apt), and we should try to keep instructions distribution agnostic if possible.
> * The current `installation.rst` is short and can be included in another document without making it overblown.
> 
> In essence, a single document that:
> 
> 1. Details installation procedure via Git repo, pip, and containers.
> 2. How to run the bot and define settings and plugins through an `entrypoint` file.
> 3. A link to `plugins.rst` for writing custom plugins.

Sounds good.

File: docs/getting_started.rst
Comment (ttuffin) [inline docs/getting_started.rst:61]: Not having a fun time with rst :disappointed:. Whats the syntax to link to an internal page?

Comment (ttuffin) [reply docs/getting_started.rst:61]: I realised this was just Github failing to render the file properly. It works in readthedocs.

File: docs/getting_started.rst
Comment (jneeven) [inline docs/getting_started.rst:125]: Can someone have a look at this? I have zero Django experience and this example probably doesn't work, but I'm not sure how Django works in the first place.

**Comment by ttuffin** [comment]
Almost there :+1: 
In regards to `deploy.rst`, the new doc (`getting-started.rst`) already covers installation via pip, repo and container. Using `supervisor` to run the bot seems like a corner-case to me and not necessary to include in docs.

**Review by ttuffin — COMMENTED** [review]
@jneeven @unode feel free to add your e-mails if you wish.

File: AUTHORS.rst
Comment (attzonko) [inline AUTHORS.rst:5]: I don't like the `Development Lead` bucket at all. I kept it from the previous author. I would be more than happy if we just have an `Active Contributors`  and `Past Contributors`

**Comment by attzonko** [comment]
By the way thanks for all the hard work on updating the docs. Looks like the TODO list is done, minus the migration docs. I am not sure we need to wait for those before merging. What do you all think?

File: README.md
Comment (attzonko) [inline README.md:89]: SSL_VERIFY = 443?

File: docs/getting_started.rst
Comment (attzonko) [reply docs/getting_started.rst:125]: I look at this but don't really know what I am looking at 😸 Honestly I am not sure we need to include the Django integration section. What is the actual use case for that?

File: docs/testing.rst
Comment (attzonko) [inline docs/testing.rst:4]: The info here seems to be duplicated from contributing.rst do we really need it in two places?

File: AUTHORS.rst
Comment (ttuffin) [reply AUTHORS.rst]: Done!

File: README.md
Comment (ttuffin) [reply README.md]: Oops, good catch. Corrected!

File: docs/testing.rst
Comment (ttuffin) [reply docs/testing.rst]: Yep you are correct. That was one of the files on the list to remove. I've merged any recent changes into `contributing.rst` and removed `testing.rst` now.

File: docs/getting_started.rst
Comment (jneeven) [reply docs/getting_started.rst:125]: Yeah fair point, I wouldn't know either... I'll remove it altogether 👍🏼

**Review by jneeven — APPROVED** [review]
LGTM! @ttuffin thanks for taking care of this

File: AUTHORS.rst
Comment (unode) [inline AUTHORS.rst:9]:
```suggestion
* Renato Alves <alves.rjc@gmail.com>
```

**Review by attzonko — APPROVED** [review]

**Comment by attzonko** [comment]
@unode Should we wait for the Migration doc or merge this and open a new one specifically for migration?

**Comment by unode** [comment]
Please go ahead and merge. I don't think we need to wait on this for the following steps.

**Pull request merged**
